### PR TITLE
Homework 4 submission

### PR DESCRIPTION
**Pull request by aniwin**
Hi, this is the homework 4 submission.

### CODE REVIEW

**Comment by CHA** [comment]
Well received, thanks

File: Homework_Week_4/task_1.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+Git workflow fundamentals
```
Comment (CHA) [inline Homework_Week_4/task_1.md:1]: ✅ Correct answers for git terms and commands

File: Homework_Week_4/task_2.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        else:
+            print("Invalid pin")
+            attempt += 1
+    print("Thank you for using the ATM.")
+
+def main_menu():
+    print("Welcome to the atm!")
+    log_in()
+
+main_menu()
```
Comment (CHA) [inline Homework_Week_4/task_2.py:38]: ⚠️ (-4) well done on completing the program. I like your approach to dividing it into multiple functions, great job.
However, to make this better, you should exclude the cash_withdrawal() method from the 'login' method.
As a general principle, one method should only do one thing. In this case your `login` method has too many responsiblity i.e. handle login and it also handle `cash withdrawal`

Take a look on the function design below:

```python
def verify_pin(pin):
   # your logic

def login(pin):
   # your logic, to verify including the retry 3 times

def cash_withdraw(balance, amount): 
   # your logic to withdraw


def atm():
   if login(pin):
      cash_withdraw()
   else
       exit()

if __name__ == '__main__':
    atm() # run the atm
```

File: Homework_Week_4/task_3_testing/atm_tests.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import unittest
```
Comment (CHA) [inline Homework_Week_4/task_3_testing/atm_tests.py:1]: ⚠️ (-3) well done on completing the unit tests. You managed to test the base case.
Would be good if you could add another test to test the edge cases i.e.
- test case where failed to input more than 3 times
- where user input less than 4 digits
- when user tries to withdraw more than the balance

**Comment by CHA** [comment]
Well done for completing the ATM program together with the unit tests. A couple of feedback:
- Instead of throwing the general Exception, try to raise a more specific error. You can create custom error specific to your business case. For example `InsufficientFundError` when user balance is less than 0
- Try to refactor your code into a smaller function. One function should have one task.
- Try to cover edge cases in unit tests.

Grade: 93
